PEP 387: Add Soft Deprecation section (#3182)
python/peps: pep-0387.txt

  platforms).

Soft Deprecation
================

> A soft deprecation can be used when using an API must no longer be used

Reviewer: ```suggestion
A soft deprecation can be used when using an API which must no longer be used
```

> A soft deprecation can be used when using an API must no longer be used

Reviewer 2: - "must" implies commanding users that they are not allowed to write it in new code, even if they really want to
-  overall context is "no longer recommended, but perfectly fine to use in existing code as long as you don't expect new features"

IMHO this feels like a bit of a tonal oddity. Is this soft-deprecated functionality an actual concrete problem for new users who aren't interested in more features but just happen to really like this module anyway?

Maybe "should" is a better word to use here?

Reviewer 2: For example optparse is mentioned down below, but not everyone considers that argparse is a proper replacement for optparse. Some people still like optparse, or just feel like they cannot use argparse without breaking their command-line parsing contract due to argparse's quirky design.

Reviewer 2: So a Soft Deprecation is intended exclusively for cases where the code in question is objectively bad and anyone who uses it is assumed to be buggy then?

e.g. it cannot be used for getopt, because getopt is the correct (and only stdlib) way to do "parse arguments according to the high-level meta-conceptual semantics POSIX describes". Many people just prefer to use alternative argument resolution strategies when faced with ambiguity.

In that case, can it be clarified to point out that Soft Deprecation must not be used for code which only fell out of *popularity*?

e.g. don't mention:
> The API remains documented and tested, but will not be developed further (no enhancement).

because the lack of enhancements is non-topical. Instead mention:
> The API remains documented and tested, but has outstanding issues which will not be fixed.

Author: Bugs in soft deprecated APIs are fixing and will continue to be fixed.

to write new code, but it remains safe to continue using it in existing
code. The API remains documented and tested, but will not be developed
further (no enhancement).

The documentation of a soft deprecation should explain why the API
should be avoided, and if possible propose a replacement.

> Soft deprecated functions must emit a SoftDeprecationWarning warning

Reviewer 3: What about soft deprecated *modules* (like optparse, for example)? It would seem better to issue this warning on import, and not have every individual function call issue a warning.

Author: I clarified this point.

> Soft deprecated functions must emit a SoftDeprecationWarning warning

Reviewer: ```suggestion
Soft deprecated functions must emit a ``SoftDeprecationWarning`` warning
```

(new in Python 3.13) in the `Python Development Mode
<https://docs.python.org/dev/library/devmode.html>`_ and if `Python is
built in debug mode
<https://docs.python.org/dev/using/configure.html#python-debug-build>`_:

> by default, this warning is not emitted.

Reviewer 3: Can we be explicit that `-Werror` and similar "turn warnings into error" approaches *must not* make soft deprecation warnings errors?

Author: Running ``python3.13 -X dev -Werror`` will treat SoftDeprecationWarning as any other warnings as errors. If you want to treat them differently, you should dig into the warnings documentation to fine tune them.

For example, ``python3.13 -X dev -Werror -Walways::SoftDeprecationWarning`` will treat all warnings as error, but SoftDeprecationWarnings will just be displayed.

I don't think that a PEP is the right place to document warnings usage.

Author: Technically, a warning is an exception. Raising a warning means raising an exception: that's what ``-Werror`` does.

warnings.warn() says **issue a warning**: https://docs.python.org/dev/library/warnings.html#warnings.warn

I can replace "this warning is not not emitted" with "this warning is not issued". Would it be better like that?

Changing a soft deprecation into a regular "hard" deprecation must

> follow the regular `Backwards Compatibility Rules`_.

Reviewer 3: Furthermore, a soft deprecation carries no implication that it will *ever* be turned into a hard deprecation. It's entirely acceptable for a soft deprecation to remain that way forever.

Author: I added a section to be more explicit that soft deprecation does NOT imply scheduling the removal of the API.

I wrote that converting a soft deprecation to a hard deprecation requires to follow the usual process to deprecate an API. Don't you think that it's enough to mean that a soft deprecated API can stay like that forever. If you don't that it's not enough, can you propose to rephase this paragraph?

Reviewer 3: How about this?

> If the decision is made to deprecate (in the regular sense) a feature that is currently soft deprecated, the deprecation must follow the regular backward compatibility rules (i.e., there is no exception because the feature is already soft deprecated).

The documentation of a soft deprecation should explain why the API
should be avoided, and if possible propose a replacement.

> Soft deprecation warnings are ignored by default. However, if the Python

Reviewer 3: ```suggestion
Soft deprecation warnings are not issued in a normal invocation of Python. However, if the Python
```

"Ignored by default" suggests that `-Werror` will enable them.
